stellar-contract-env-host/src/host.rs

```diff
@@ -1549,30 +1549,137 @@ impl CheckedEnv for Host {
+            .iter()
+            .map(|v| u32::try_from(*v))
+            .collect::<Result<Vec<u32>, ConversionError>>()?
+            .as_slice()
+        {
+            self.visit_obj(b, move |hv: &Vec<u8>| {
+                let end_idx = b_pos.checked_add(*len).ok_or_else(|| {
+                    HostError::WithStatus(
+                        String::from("u32 overflow"),
+                        ScStatus::HostFunctionError(ScHostFnErrorCode::InputArgsInvalid),
+                    )
+                })? as usize;
+                if end_idx > hv.len() {
+                    return Err(HostError::WithStatus(
+                        String::from("index out of bound"),
+                        ScStatus::HostObjectError(ScHostObjErrorCode::VecIndexOutOfBound),
+                    ));
+                }
+                self.with_current_frame(|frame| match frame {
+                    Frame::ContractVM(vm) => vm.with_memory_access(|mem| {
+                        Ok(mem.set(*lm_pos, &hv.as_slice()[*b_pos as usize..end_idx])?)
+                    }),
+                    Frame::HostFunction(_) => {
+                        Err(HostError::General("linear memory not supported"))
+                    }
+                    #[cfg(feature = "testutils")]
+                    Frame::TestContract(id) => {
+                        Err(HostError::General("linear memory not supported"))
+                    }
+                })
+            })?;
+            Ok(().into())
+        } else {
+            return Err(HostError::WithStatus(
+                String::from("failed conversion to u32"),
+                ScStatus::HostFunctionError(ScHostFnErrorCode::InputArgsWrongType),
+            ));
+        }
     }
 
-    fn binary_copy_from_guest_mem(
+    fn binary_copy_from_linear_memory(
         &self,
         b: Object,
-        i: RawVal,
-        j: RawVal,
-        l: RawVal,
-    ) -> Result<RawVal, HostError> {
-        todo!()
+        b_pos: RawVal,
+        lm_pos: RawVal,
+        len: RawVal,
+    ) -> Result<Object, HostError> {
+        #[cfg(not(feature = "vm"))]
+        todo!();
+        #[cfg(feature = "vm")]
+        if let [b_pos, lm_pos, len] = [b_pos, lm_pos, len]
+            .iter()
+            .map(|v| u32::try_from(*v))
+            .collect::<Result<Vec<u32>, ConversionError>>()?
+            .as_slice()
+        {
+            let mut vnew = self.visit_obj(b, |hv: &Vec<u8>| Ok(hv.clone()))?;
+            let end_idx = b_pos.checked_add(*len).ok_or_else(|| {
+                HostError::WithStatus(
+                    String::from("u32 overflow"),
+                    ScStatus::HostFunctionError(ScHostFnErrorCode::InputArgsInvalid),
+                )
+            })? as usize;
+            self.with_current_frame(|frame| match frame {
+                Frame::ContractVM(vm) => vm.with_memory_access(|mem| {
+                    /* we would potentially let the vec grow */
```

Comment:
Nitpick: this is a one-line comment so use the `//` style

Reply:
Fixed.